storage/provider/openstack.go

```diff
@@ -0,0 +1,443 @@
+}
+
+var _ storage.Provider = (*OpenstackProvider)(nil)
+
+// VolumeSource implements storage.Provider.
+func (p *OpenstackProvider) VolumeSource(environConfig *config.Config, providerConfig *storage.Config) (storage.VolumeSource, error) {
+
+	if err := p.ValidateConfig(providerConfig); err != nil {
+		return nil, err
+	}
+
+	openstackClient, err := p.ClientFactory(providerConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &openstackVolumeSource{openstackClient}, nil
+}
+
+// FilesystemSource implements storage.Provider.
+func (p *OpenstackProvider) FilesystemSource(environConfig *config.Config, providerConfig *storage.Config) (storage.FilesystemSource, error) {
+	return nil, errors.NotSupportedf("filesystems")
+}
+
+// Supports implements storage.Provider.
+func (p *OpenstackProvider) Supports(kind storage.StorageKind) bool {
+	switch kind {
+	case storage.StorageKindBlock:
+		return true
+	}
+	return false
+}
+
+// Scope implements storage.Provider.
+func (s *OpenstackProvider) Scope() storage.Scope {
+	panic("not implemented")
+}
+
+// ValidateConfig implements storage.Provider.
+func (p *OpenstackProvider) ValidateConfig(cfg *storage.Config) error {
+
+	for _, cfgKey := range []string{
+		openstackCfgKeyUrl,
+		openstackCfgKeyUser,
+		openstackCfgKeySecrets,
+		openstackCfgKeyRegion,
+		openstackCfgKeyTenantName} {
+		if _, ok := cfg.Attrs()[cfgKey]; !ok {
+			return errors.NotAssignedf("requisite configuration was not set: %v", cfgKey)
+		}
+	}
+
+	return nil
+}
+
+// Dynamic implements storage.Provider.
+func (p *OpenstackProvider) Dynamic() bool {
+	panic("not implemented")
+}
+
+type openstackVolumeSource struct {
+	openstackClient OpenstackAdapter
+}
+
+var _ storage.VolumeSource = (*openstackVolumeSource)(nil)
+
+// CreateVolumes implements storage.VolumeSource.
+func (s *openstackVolumeSource) CreateVolumes(args []storage.VolumeParams) (vols []storage.Volume, volAtt []storage.VolumeAttachment, returnErr error) {
+
+	vols = make([]storage.Volume, 0, len(args))
+	volAtt = make([]storage.VolumeAttachment, 0, len(args))
+
+	for _, a := range args {
+		if a.Provider != CinderProviderType {
+			logger.Debugf(
+				"openstack storage provider was passed a request for another provider type: %v",
+				a.Provider,
+			)
+			continue
+		}
+
+		createdVol, err := s.openstackClient.CreateVolume(a.Size, a.Tag)
+		if err != nil {
+			return nil, nil, err
+		}
+		vols = append(vols, createdVol)
+
+		// If the method exits with an error, be sure to delete any
+		// created volumes so that we're idempotent. We create several
+		// clousures instead of one to take advantage of the loop
+		// parameter.
+		defer func(arg storage.VolumeParams, createdVol storage.Volume) {
+			if returnErr == nil || createdVol == (storage.Volume{}) {
+				return
+			}
+
+			if arg.Attachment != nil {
+				volAttachments, err := s.openstackClient.ListVolumeAttachments(string(arg.Attachment.InstanceId))
+				if err != nil {
+					panic(err)
```

Comment:
Whoa! {anicking is a bit drastic, and suggests that the caller did something wrong.
Really there's nothing we can do at this point, so panicking isn't helpful. Just log the error and continue. Once harvesting is implemented we'll retry.